File: GameShelf/Services/Epic/EpicCredentials.swift
//
//  EpicCredentials.swift
//  GameShelf
//

import Foundation

/// Lo que hace falta para hablar con Epic en nombre del usuario.
struct EpicCredentials: Sendable, Equatable {
  let accessToken: String
  let refreshToken: String

  /// Cuando deja de servir el token de acceso. Epic da unas ocho horas.
  let expiresAt: Date

  /// Cuando deja de servir el de refresco. Epic si manda este dato.
  var refreshExpiresAt: Date?

  /// Identificador de la cuenta, que hace falta para pedir la biblioteca.
  var accountID: String?

  /// Nombre visible de la cuenta. Solo para mostrarlo.
  var displayName: String?

  /// Si el token de acceso ya no sirve, o le queda muy poco.
  ///
  /// Se adelanta un minuto, igual que en PSN: pedir con un token que expira
  /// mientras la peticion va en camino da un 401 evitable.
  func isExpired(at fecha: Date = Date(), margen: TimeInterval = 60) -> Bool {
    fecha.addingTimeInterval(margen) >= expiresAt
  }
}

/// Por que fallo la autenticacion con Epic.
///
/// Los codigos vienen de la respuesta real y se comprobaron contra la API.
enum EpicAuthError: LocalizedError, Equatable {
  /// El codigo no existe o ya se uso. Duran muy poco.
  case codigoInvalido

  /// El refresh token ya no sirve: toca copiar un codigo nuevo.
  case sesionExpirada

  /// Epic respondio algo que no se esperaba.
  case respuestaInesperada(String)

  /// No hay credenciales guardadas todavia.
  case sinCredenciales

  var errorDescription: String? {
    switch self {
    case .codigoInvalido:
      String(
        localized: "El codigo de Epic no sirve o ya se uso.",
        comment: "Error de autenticacion con Epic"
      )
    case .sesionExpirada:
      String(localized: "La sesion de Epic caduco.", comment: "Error de autenticacion con Epic")
    case .respuestaInesperada(let detalle):
      String(
        localized: "Epic respondio algo inesperado: \(detalle)",
        comment: "Error de autenticacion con Epic"
      )
    case .sinCredenciales:
      String(
        localized: "Todavia no has conectado tu cuenta de Epic.",
        comment: "Error de autenticacion con Epic"
      )
    }
  }

  var recoverySuggestion: String? {
    switch self {
    case .codigoInvalido:
      // Los codigos de Epic caducan en segundos: lo mas comun es tardar entre
      // copiarlo y pegarlo, no que este mal escrito.
      String(
        localized: "Los codigos caducan enseguida. Genera uno nuevo y pegalo sin demorarte.",
        comment: "Como recuperarse de un codigo de Epic invalido"
      )
    case .sesionExpirada:
      String(
        localized: "Vuelve a generar el codigo desde el navegador y pegalo aca.",
        comment: "Como recuperarse de una sesion de Epic caducada"
      )
    case .sinCredenciales:
      String(localized: "Conecta tu cuenta desde Ajustes.", comment: "Como conectar PSN")
    case .respuestaInesperada:
      nil
    }
  }

  /// Si la unica salida es que el usuario genere un codigo nuevo.
  var necesitaCodigoNuevo: Bool {
    self == .codigoInvalido || self == .sesionExpirada || self == .sinCredenciales
  }
}

File: android/app/src/main/java/com/gameshelf/data/epic/EpicCredentials.kt
package com.gameshelf.data.epic

import android.content.Context
import com.gameshelf.R
import com.gameshelf.data.net.UserFacingError
import java.time.Duration
import java.time.Instant

/** Lo que hace falta para hablar con Epic en nombre del usuario. */
data class EpicCredentials(
  val accessToken: String,
  val refreshToken: String,

  /** Cuando deja de servir el token de acceso. Epic da unas ocho horas. */
  val expiresAt: Instant,

  /** Cuando deja de servir el de refresco. Epic si manda este dato. */
  val refreshExpiresAt: Instant? = null,

  /** Identificador de la cuenta, que hace falta para pedir la biblioteca. */
  val accountID: String? = null,

  /** Nombre visible de la cuenta. Solo para mostrarlo. */
  val displayName: String? = null,
) {
  /**
   * Si el token de acceso ya no sirve, o le queda muy poco.
   *
   * Se adelanta un minuto, igual que en PSN: pedir con un token que expira
   * mientras la peticion va en camino da un 401 evitable.
   */
  fun isExpired(at: Instant = Instant.now(), margen: Duration = Duration.ofSeconds(60)): Boolean =
    !at.plus(margen).isBefore(expiresAt)
}

/**
 * Por que fallo la autenticacion con Epic.
 *
 * Los codigos vienen de la respuesta real y se comprobaron contra la API.
 */
sealed class EpicAuthError : Exception(), UserFacingError {

  /** El codigo no existe o ya se uso. Duran muy poco. */
  data object CodigoInvalido : EpicAuthError()

  /** El refresh token ya no sirve: toca copiar un codigo nuevo. */
  data object SesionExpirada : EpicAuthError()

  /** Epic respondio algo que no se esperaba. */
  data class RespuestaInesperada(val detalle: String) : EpicAuthError()

  /** No hay credenciales guardadas todavia. */
  data object SinCredenciales : EpicAuthError()

  override fun message(context: Context): String = when (this) {
    CodigoInvalido -> context.getString(R.string.epic_error_code_invalid)
    SesionExpirada -> context.getString(R.string.epic_error_session_expired)
    is RespuestaInesperada -> context.getString(R.string.epic_error_unexpected, detalle)
    SinCredenciales -> context.getString(R.string.epic_error_not_connected)
  }

  override fun recovery(context: Context): String? = when (this) {
    // Los codigos de Epic caducan en segundos: lo mas comun es tardar entre
    // copiarlo y pegarlo, no que este mal escrito.
    CodigoInvalido -> context.getString(R.string.epic_recovery_code_expires_fast)
    SesionExpirada -> context.getString(R.string.epic_recovery_generate_again)
    SinCredenciales -> context.getString(R.string.recovery_connect_from_settings)
    is RespuestaInesperada -> null
  }

  /** Si la unica salida es que el usuario genere un codigo nuevo. */
  val necesitaCodigoNuevo: Boolean
    get() = this is CodigoInvalido || this is SesionExpirada || this is SinCredenciales

  override val message: String
    get() = when (this) {
      CodigoInvalido -> "Codigo de Epic invalido"
      SesionExpirada -> "Sesion de Epic expirada"
      is RespuestaInesperada -> "Respuesta inesperada de Epic: $detalle"
      SinCredenciales -> "Sin credenciales de Epic"
    }
}
